### PR TITLE
Lo fine tune

### PR DESCRIPTION
**Pull request by ik1xpv**
Tuning of LO to 1Hz sems ok to me.  Hayati code does all the job THANKS !
To test it you have to enable LO fine tuning in the ExtIO dialog.  This setting parameter is not yet saved. you have to enable it at HDSDR start.
I plan to rebase to the master before going on. 
I selected just one of the faster Hayati's solutions. A better selection strategy may be added. My CPU load increases by 1/10th.

### CODE REVIEW

File: ExtIO_sddc/r2iq.cpp [hunk 1]
```diff
@@ -92,9 +92,9 @@ int r2iqControlClass::Setdecimate(int dec)
 	return mratio[mdecimation];
 }
 
-void r2iqControlClass::Updt_SR_LO_TUNE(int srate_idx, int64_t* fLO, int64_t* fTune)
+void r2iqControlClass::Updt_SR_LO_TUNE(int srate_idx, int64_t* pfLO, int64_t* pfTune)
```
Comment (howard0su) [inline ExtIO_sddc/r2iq.cpp:95]: this function and UptTunerFrq is so similar. can we merge them?

File: ExtIO_sddc/radio/HF103Radio.cpp [hunk 2]
```diff
@@ -27,7 +27,7 @@ bool HF103Radio::UpdateattRF(int att)
 {
     if (att > 31) att = 31;
     if (att < 0) att = 0;
-    uint8_t d = (31 - att) << 1; // bit0 =0
+    uint8_t d = (31 - att) << 1; // bit0 = 0
```
Comment (howard0su) [inline ExtIO_sddc/radio/HF103Radio.cpp:30]: unrelated space change?

File: ExtIO_sddc/RadioHandler.cpp [hunk 3]
```diff
@@ -121,6 +122,7 @@ void RadioHandlerClass::AdcSamplesProcess()
 		{
 			if (rd == 1)
 			{
+				FineTuneLO((complexf*) obuffers[idx], EXT_BLOCKLEN, rd);
```
Comment (howard0su) [inline ExtIO_sddc/RadioHandler.cpp:125]: Shall we do it in r2iq thread instead of here? this thread is heavy loaded now and it can slow down the usb polling.

**Review by howard0su — CHANGES_REQUESTED** [review]

File: ExtIO_sddc/r2iq.cpp (hunk 1 above)
Comment (howard0su) [reply ExtIO_sddc/r2iq.cpp:95]: actually, I think this function should not belong to r2iq. All r2iq needs is just a tunebin offset. Maybe this logic should be in RadioHandler and set tubebin to r2iq.

**Comment by howard0su** [comment]
#20 touched some part of r2iq code. be aware of merge conflict.

File: ExtIO_sddc/r2iq.cpp (hunk 1 above)
Comment (ik1xpv) [reply ExtIO_sddc/r2iq.cpp:95]: Yes your right Todo

File: ExtIO_sddc/radio/HF103Radio.cpp (hunk 2 above)
Comment (ik1xpv) [reply ExtIO_sddc/radio/HF103Radio.cpp:30]: Yes

File: ExtIO_sddc/r2iq.cpp (hunk 1 above)
Comment (ik1xpv) [reply ExtIO_sddc/r2iq.cpp]: > this function and UptTunerFrq is so similar. can we merge them?
Done. Only UptTunerFrq now.

File: ExtIO_sddc/ExtIO_sddc.cpp [hunk 4]
```diff
@@ -404,16 +410,21 @@ extern "C"
 long EXTIO_API GetHWLO(void)
 {
     EnterFunction();
-
-	return (long)(glLOfreq & 0xFFFFFFFF);
+	if (RadioHandler.GetFine_LO())
```
Comment (howard0su) [inline ExtIO_sddc/ExtIO_sddc.cpp:413]: why not make glLocoor == 0 in the other case, then we can normalize the code here just return glLOfreq + glLOcorr

File: ExtIO_sddc/ExtIO_sddc.cpp [hunk 5]
```diff
@@ -620,7 +631,8 @@ int  EXTIO_API ExtIoSetSrate(int srate_idx)
 		giExtSrateIdx = srate_idx;
 		gExtSampleRate = (unsigned)(newSrate + 0.5);      // update
 
-		r2iqCntrl.Updt_SR_LO_TUNE(srate_idx, &glLOfreq, &glTunefreq);
+		r2iqCntrl.Setdecimate(4 - srate_idx); // reverse order 32,16,8,4,2 MHz
```
Comment (howard0su) [inline ExtIO_sddc/ExtIO_sddc.cpp:634]: I think this is not necessary. but I can be wrong and need to test

File: ExtIO_sddc/ExtIO_sddc.vcxproj [hunk 6]
```diff
@@ -254,6 +254,8 @@
     </Link>
   </ItemDefinitionGroup>
   <ItemGroup>
+    <ClInclude Include="pffft\fmv.h" />
```
Comment (howard0su) [inline ExtIO_sddc/ExtIO_sddc.vcxproj:257]: please update CMakeFile.txt to add pf_mixer.cpp to SRCS list

Comment (howard0su) [reply ExtIO_sddc/ExtIO_sddc.vcxproj:257]: and add include folder as well. I can help you want to.

File: ExtIO_sddc/RadioHandler.cpp [hunk 7]
```diff
@@ -453,3 +467,18 @@ void RadioHandlerClass::UpdBiasT_VHF(bool flag)
 		hardware->FX3UnsetGPIO(BIAS_VHF);
 }
 
+void RadioHandlerClass::FineTuneLO(complexf* input, int nsample, int rd)
+{
+	if (GetFine_LO() && (glLOcorr != 0.0F)) // if Fine Local Oscillator Tuning allowed and required
```
Comment (howard0su) [inline ExtIO_sddc/RadioHandler.cpp:472]: why not preventing set glOcrr to non-zero when FineLocalOSC is disabled? This will reduce the checks in couple of places.

Comment (ik1xpv) [reply ExtIO_sddc/RadioHandler.cpp:472]: You are right as usual :-)

File: ExtIO_sddc/ExtIO_sddc.cpp (hunk 5 above)
Comment (ik1xpv) [reply ExtIO_sddc/ExtIO_sddc.cpp:634]: I left SetDecimate and tested it ok.

File: ExtIO_sddc/ExtIO_sddc.cpp (hunk 4 above)
Comment (ik1xpv) [reply ExtIO_sddc/ExtIO_sddc.cpp:413]: Ok done

File: ExtIO_sddc/RadioHandler.cpp (hunk 3 above)
Comment (ik1xpv) [reply ExtIO_sddc/RadioHandler.cpp:125]: I have to think about. Here is applied to all the EXT_BLOCKLEN buffer.

Comment (howard0su) [reply ExtIO_sddc/RadioHandler.cpp:125]: I see. maybe we let it be as the first version.

I am thinking to move the callback of EXT_BLOCK to separate threads. We can do the optimization later.

File: ExtIO_sddc/ExtIO_sddc.vcxproj (hunk 6 above)
Comment (ik1xpv) [reply ExtIO_sddc/ExtIO_sddc.vcxproj:257]: in my ExtIO_sddc/ExtIO_sddc.vcxproj I see:
 </ItemDefinitionGroup>
  <ItemGroup>
    <ClInclude Include="pffft\fmv.h" />

I have only  CMakeLists.txt no CMakeFile.txt ?

File: ExtIO_sddc/RadioHandler.cpp (hunk 3 above)
Comment (ik1xpv) [reply ExtIO_sddc/RadioHandler.cpp:125]: Ok

File: ExtIO_sddc/ExtIO_sddc.vcxproj (hunk 6 above)
Comment (howard0su) [reply ExtIO_sddc/ExtIO_sddc.vcxproj:257]: my bad. it should be CMakeLists.txt.

Comment (howard0su) [reply ExtIO_sddc/ExtIO_sddc.vcxproj:257]: Apply the following patch to CMakeLists.txt:

diff --git a/ExtIO_sddc/CMakeLists.txt b/ExtIO_sddc/CMakeLists.txt
index f008757..e500608 100644
--- a/ExtIO_sddc/CMakeLists.txt
+++ b/ExtIO_sddc/CMakeLists.txt
@@ -3,9 +3,9 @@ project(ExtIO_RX888)

 set(CMAKE_CXX_FLAGS "-O2 /std:c++17 /EHsc /W3")

-file(GLOB SRC "*.cpp" "radio/*.cpp" "CyAPI/*.cpp" extio.def resource.rc manifest.xml)
+file(GLOB SRC "*.cpp" "radio/*.cpp" "CyAPI/*.cpp" "pffft/*.cpp" extio.def resource.rc manifest.xml)

-include_directories("." "CyAPI/" "FFTW3/")
+include_directories("." "CyAPI/" "FFTW3/" "pffft/")

 set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -D_DEBUG")
